rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+2. [Test those changes in isolation][reference-module-testing]
+3. Deploy those changes to their team's development/staging environment to verify their impact
+4. Deploy those changes to production once they are tested and approved
+
+Scalr will be an integral part of each step in this workflow, as detailed below.
+
+#### Scalr Workflow for Module Testing
+[scalr-workflow-module-testing]: #scalr-workflow-for-module-testing
+
+Each testing module will have a CLI-driven Scalr workspace with a remote operation backend that runs all `terraform` commands via Scalr. `terraform` commands will be run manually by the engineer, or possibly as part of a GitHub Actions workflow. Scalr's remote operation backend means the actual work will be done remotely by Scalr, using [Scalr's AWS permissions][reference-permissions], and the `terraform` CLI will show the run output locally.
+
+##### Continuous Integration
+
+- Engineer creates or modifies module configuration locally
+- Engineer pushes changes to GitHub branch
+- GitHub Actions workflow runs:
+  - `terraform validate`
+  - `terraform fmt`
+- Engineer (or GitHub Actions) tests module creation/destruction via Scalr (see workflow description in [Module Testing Workflow section][reference-module-testing])
+- Engineer opens pull request
+
+##### Code Review/Approval
+
+- Review &amp; approval handled via GitHub PR
+- Branch protection rules ensure that PR cannot be merged without approval
+
+##### Applying Changes
+
+- Engineer merges approved PR
+- If module versioning is employed, GitHub Actions triggers workflow to publish new version of module to module registry
+
+Once a new version of the module is published, the module version can be updated in the development/staging and production contexts using the workflows described below.
+
+If module versioning is not employed, changes to the module will need to be applied immediately in any dependent root modules. In the development/staging context, this can be handled automatically by configuring [run triggers](https://docs.scalr.com/en/latest/workspaces.html#run-triggers) in the Scalr workspace. In the production context, however, there will be an extra layer of protection to ensure that changes don't impact production before they're ready to be applied there. This is described in more detail in the [production workflow section][scalr-workflow-production] below.
+
+#### Scalr Workflow for Applying Changes In Development/Staging Context
+[scalr-workflow-development-staging]: #scalr-workflow-for-applying-changes-in-developmentstaging-context
+
+Each root module will have a GitHub-integrated Scalr workspace. Scalr will automatically run `terraform plan` when a pull request is opened, and attempt to automatically run `terraform apply` on merge. It will also be possible to run `terraform plan` locally using Scalr's remote operation backend, but not `terraform apply`; this is to ensure that GitHub acts as the source of truth for controlling Terraform state.
```

Comment:
is this a requirement for dev environments as well? having to go through pull requests to run an apply feels a bit cumbersome to me.

Reply:
This is a technical limitation of GitHub-integrated Scalr workspaces, so it will apply to both dev and prod. That said, the expectation is that any changes will have to be reviewed before they're applied, which aligns with our current engineering and devops workflows. Let me see if I can make that more explicit.

Reply:
Added some clarification: https://github.com/mbta/technology-docs/pull/16/commits/de7cdb955da96edb4d5f32ac96d941e07a6ca65f